job-exec: minimal support for multiuser exec

grondo opened the pull request:
This PR tweaks the job-exec 'exec' implementation to attempt to run the IMP for any multiuser job requests, i.e. those where job->userid != getuid().

The `exec.imp` config parameter must be set to a path of a `flux-imp` installed setuid for multiuser to be enabled. Currently, if the IMP isn't configured, multi-user job requests get an exec exception at start. `exec.imp` can be overridden by `imp=` on the module load cmdline.

This PR implments the bare minimum to get this working. We need to re-architect the  job-exec module soon anyway, so it wouldn't be prudent to spend extra time with an optimal integration.

The PR is still a WIP so I can add at least *some* testing.

Edit: this is built on top of #2821 

Also edit: the current work here doesn't check if the `flux_imp_path` is actually installed setuid. This could be done, but it will actually be handy for testing purposes not to require it (for `make check` we can replace the imp with a small shell script, mwahaha)

garlick commented on src/modules/job-exec/job-exec.c:
```diff
@@ -729,6 +729,11 @@ static void jobinfo_start_continue (flux_future_t *f, void *arg)
         goto done;
     }
+    if (job->multiuser
+        && !(job->J = strdup (jobinfo_kvs_lookup_get (f, "J")))) {
```
I couldn't find anything that would indicate `strdup (NULL)` is required to return NULL, which I think is the assumption here.  Might want to handle `strdup()` and `jobinfo_kvs_lookup_get()` failures separately.

grondo replied:
Hm, there might be other problems in this function. Other fields like `jobspec` and `R` are not copied, but afaics the future containing these results is destroyed at function exit. I think so far no crashes because maybe the data is only accessed while the function is still running.

Probably the easiest thing to do here is to save the future until data isn't needed anymore.

garlick commented on src/modules/job-exec/job-exec.c:
```diff
@@ -836,6 +841,11 @@ static flux_future_t *jobinfo_start_init (struct jobinfo *job)
+        && (!(f_kvs = flux_jobid_kvs_lookup (h, job->id, 0, "J"))
+        || flux_future_push (f, "J", f_kvs) < 0)) {
+        goto err;
+    }
```
can this code leak `f_kvs` on error?

grondo replied:
yes, and the other calls in this function seem to suffer the same error. Nice catch!

garlick commented on src/modules/job-exec/job-exec.c:
```diff
@@ -873,6 +883,10 @@ static int job_start (struct job_exec_ctx *ctx, const flux_msg_t *msg)
+
+    if (job->userid != getuid ())
+        job->multiuser = 1;
+
```
Just noticed we have mixed use of `getuid()` and `geteuid()` to get the instance owner throughout the code base.  Since we're not manipulating privilege in flux-core, it seems like they are equivalent.  Do you concur?  (We should probably strive for consistency, but for another day mabye)

grondo replied:
I can't think of a reason why the real and effective uid would ever be different in the broker, except by accident. Given that, wouldn't calling `geteuid()` imply that this *could* be the case, and cause confusion (as it perhaps has here). Or is `geteuid()` more correct in this case?

garlick replied:
I  agree that `getuid()` is less laden with implication and so probably more appropriate.  I'll open an issue on the consistency.

garlick commented:
I was about to resolve new conflicts between this PR and #2821 so I could give it a test.  If you already have that resolved, feel free to push here and save me duplicating effort :-)

grondo commented:
> I was about to resolve new conflicts between this PR and #2821 so I could give it a test. If you already have that resolved, feel free to push here and save me duplicating effort :-)

Sorry, I was driving a kid to school.

The "rebase" was just to drop all but the last two commits (and rebase on top of #2821)

garlick commented:
Sorry I didn't look too closely at what was failing.  Spinning up a test.

garlick commented:
Manual multiuser testing with #2822,  #2821, and #2820 applies on master is looking good!

I created `/usr/local/etc/flux/conf.d/exec.toml` containing
```toml
[exec]

imp = "/usr/local/libexec/flux/flux-imp"
```
On startup, in the dmesg I see:
```
2020-03-12T17:48:44.527660Z job-exec.debug[0]: using default shell path /usr/local/libexec/flux/flux-shell
2020-03-12T17:48:44.527683Z job-exec.debug[0]: using imp path /usr/local/libexec/flux/flux-imp
```
Yay!  Can I run jobs as me?  Yes!
```
$ flux mini run id
uid=5588(garlick) gid=5588(garlick) groups=5588(garlick),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),116(lpadmin),122(sambashare)
```

grondo commented:
You gave away your secret uid!

garlick commented:
Well at least I didn't disclose the root uid!

grondo commented:
C'mon we don't even *joke* about that!

grondo commented:
Forced a push with a couple fixes for comments. Note this is still WIP until I can get some basic sanity tests in.

grondo commented:
Ok, I've made some updates here to get multiuser kill working. This won't pass in Travis until we've tagged a version of flux-security that can be built in the docker image.

To test multi-user exec and kill, there's some new work to add Python bindings for `flux_sign_wrap_as(3)`, and a test script `t/scripts/sign-as.py` to allow sign-wrapping an arbitrary payload as any userid using the `none` mechanism.

The tests then can submit a *J* that appears to be from a different userid, activating the multiuser branches. A fake `imp.sh` replaces the normal IMP to make sure the execution system ran the flux-imp helper for the multiuser job case.

grondo commented:
Ok, I've added a change here to update flux-security in our docker image to v0.4.0, which includes the required `flux_sign_wrap_as(3)`. This PR could now use a review.

Let me just sanity check a version with installed, setuid IMP, and also try to fix the LGTM unused import warning.

garlick commented:
Looks like one builder failed here:
```
not ok 4 - job-exec: kill multiuser job uses the IMP
FAIL: t2404-job-exec-multiuser.t 4 - job-exec: kill multiuser job uses the IMP
#	
#		FAKE_USERID=42 &&
#		flux mini run --dry-run -n2 -N2 sleep 1000 | \
#		    flux python ${SIGN_AS} ${FAKE_USERID} > sleep-job.signed &&
#		id=$(FLUX_HANDLE_USERID=${FAKE_USERID} \
#		    flux job submit --flags=signed sleep-job.signed) &&
#		flux job list-ids ${id} > ${id}.json &&
#		jq -e ".userid == 42" < ${id}.json &&
#		flux job cancel ${id} &&
#		test_expect_code 143 run_timeout 30 flux job status -v ${id} &&
#		flux dmesg | grep "test-imp: Kill .*signal 15"
#	
# failed 1 among 4 test(s)
1..4
ERROR: t2404-job-exec-multiuser.t - exited with status 1

```

grondo commented:
I'll look into it, some kind of race in the exit code of the test:
```
flux-job: 18371051520: exception type=cancel
test_expect_code: command exited with 1, we wanted 143 run_timeout 30 flux job status -v 18371051520
not ok 4 - job-exec: kill multiuser job uses the IMP
```

I can't reproduce this locally and when I restarted the builder all tests passed. I'll keep trying.

grondo commented:
This time got a hang. Looks like `t4000-issues-test-driver.t` was the hanging test. Haven't seen that one before.

grondo commented:
Ok, race was obvious - - if a cancel comes in during shell.init phase then the exit code is 1, if after shell.init, then the tasks are sent SIGTERM and the exit code is 143.

Fixed, squashed and force-pushed.

grondo commented:
Hit a hang in ASan builder, the culprit appeared to be `t2609-job-shell-events.t`. Just restarted the builder for now.

grondo commented:
Oh, I also just realized the `configure` args for flux-security install will need to be modified also. I'll roll that into the existing `docker:` commit.

grondo commented:
Another change I've added here. In the travis builders that auto-deploy docker images, build with `configure` args suitable for a system installed flux, i.e. `--prefix=/usr --sysconfdir=/etc --with-systemdsystemunitdir=/etc/systemd/system`.

grondo commented:
Experimentally switched the address-sanitizer check to a Github workflow action instead of travis-ci, so see if it becomes a little more reliable.

grondo commented:
That seemed to work so I've updated the mergify config.

I think that means this PR will have to be manually merged.

grondo commented:
Ok, added a 40m timeout to the github asan action since github workflows have a lot longer timeout that travis. So far the ASan tests seem to take about 25m under GitHub -- not the improvement I was looking for, but a little better...

garlick commented:
Shall I push the button?

grondo commented:
Sure!

On Sun, Mar 22, 2020, 4:15 PM Jim Garlick <notifications@github.com> wrote:

> Shall I push the button?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2822#issuecomment-602292023>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFVEUQDUIRD6DG62UPK4PDRI2LZJANCNFSM4LGEU57A>
> .
>

grondo commented:
Thanks!
